lib/saluki-env/src/autodiscovery/providers/remote_agent.rs

```diff
@@ -0,0 +1,128 @@
+use std::sync::{Arc, Mutex};
+
+use async_trait::async_trait;
+use futures::StreamExt;
+use saluki_error::GenericError;
+use tokio::sync::mpsc::{self, error::TrySendError};
+use tracing::{debug, error, info, warn};
+
+use crate::autodiscovery::{AutoDiscoveryProvider, AutodiscoveryEvent, Config};
+use crate::helpers::remote_agent::RemoteAgentClient;
+
+/// An autodiscovery provider that uses the Datadog Agent's internal gRPC API to receive autodiscovery updates.
+pub struct RemoteAgentAutoDiscoveryProvider {
+    client: RemoteAgentClient,
+    event_senders: Arc<Mutex<Vec<mpsc::Sender<AutodiscoveryEvent>>>>,
+    listener_started: bool,
+}
+
+impl RemoteAgentAutoDiscoveryProvider {
+    /// Creates a new `RemoteAgentAutoDiscoveryProvider` that uses the remote client to receive autodiscovery updates.
+    pub fn new(client: RemoteAgentClient) -> Self {
+        Self {
+            client,
+            event_senders: Arc::new(Mutex::new(Vec::new())),
+            listener_started: false,
+        }
+    }
+
+    async fn start_background_listener(&mut self) -> Result<(), GenericError> {
+        if self.event_senders.lock().unwrap().is_empty() {
+            return Err(GenericError::msg(
+                "Cannot start background listener without subscribers",
+            ));
+        }
+
+        if self.listener_started {
+            return Ok(());
+        }
+
+        debug!("Starting autodiscovery background listener");
+
+        let mut client = self.client.clone();
+        let event_senders = self.event_senders.clone();
+
+        tokio::spawn(async move {
+            let mut autodiscovery_stream = client.get_autodiscovery_stream();
+
+            info!("Autodiscovery stream established");
+
+            while let Some(result) = autodiscovery_stream.next().await {
+                match result {
+                    Ok(response) => {
+                        for proto_config in response.configs {
+                            let config = Config::from(proto_config.clone());
+                            let event = AutodiscoveryEvent { config };
+
+                            broadcast_event(&event_senders, event).await;
+                        }
+                    }
+                    Err(status) => {
+                        error!("Error from autodiscovery stream: {}", status);
+                        break;
+                    }
+                }
+            }
+
+            warn!("Autodiscovery stream closed");
+        });
+
+        self.listener_started = true;
+        Ok(())
+    }
+}
+
+async fn broadcast_event(senders: &Arc<Mutex<Vec<mpsc::Sender<AutodiscoveryEvent>>>>, event: AutodiscoveryEvent) {
```

Comment:
Related to the questions below: we should just use `tokio::sync::broadcast`, which already solves this in a cleaner way.